Enable simultaneous control of multiple Feedback360 servos

Yupopyoi opened the pull request:
Fix  #1 :  **Enable simultaneous control of multiple Feedback360 servos**

- Refactored src/FeedBackServo.h and .cpp to support multiple motors
  - Removed static declarations of angle and other variables that should be instance-specific
- Introduced non-blocking P-control using Update() to be called from loop()
- This allows parallel-like control of multiple servos without blocking
- Added example sketch (MultiMotors.ino) demonstrating concurrent control
- Verified parallel control using Arduino UNO and two Feedback360 servos

**複数のFeedback360サーボの同時制御**

- src内の.hおよび.cppファイルに変更を加え、複数モーターの同時制御を可能にしました
  - モーターごとに異なる値を持つ angle などの変数が static だったため、複数制御ができなかったと考えられます
- loop() 内で Update() を呼び出す非ブロッキング型にし、擬似並列制御を可能にしました
- 複数台同時P制御の挙動を確認できる MultiMotors.ino を追加しました
- Arduino UNO + モーター2台で動作確認
- 既存のサンプル（Read.ino と Rotate.ino）は、新しいAPIでは動作しなくなっています

HyodaKazuaki reviewed (commented):
Thank you for your pull request. These are great changes and almost ready to merge.
It seems these changes are breaking changes and I'd like to release this in the next major version update.
Also, thank you for aligning the member function names to Upper Camel Case.
Due to my oversight, the naming convention deviates from Arduino's standard.
I'll create a new issue to address this.
I have one question, if you don't mind answering it.

PRを送っていただきありがとうございます。とても良いと感じました。ほぼOKです。
このPRは破壊的変更なので、次回のメジャーバージョンでの更新でリリースしたいと思います。
メンバー変数名もUpper Camel Caseで揃えていただきありがとうございます。
私の不手際により、命名規則がArduinoのものからずれてしまっています。
これは別のIssueに分けて対応したいと思います。
1つ質問があるので、よろしければ答えていただけると嬉しいです。

Yupopyoi commented on src/FeedBackServo.h:
```diff
@@ -1,34 +1,51 @@
     static const int q2min;
     static const int q3max;
+
+    static FeedBackServo* instances[6];
```
I should have left an explanation there. Sorry about that.

The reason why the instances array is limited to 6 is because the board I'm primarily using is the Arduino Mega, which has 6 external interrupt pins. So, this effectively limits the number of feedback servos that can be handled simultaneously to 6.

If using an Arduino Uno, for example, which only has 2 external interrupt pins, then an array of size 2 would suffice.

It might be a good idea in the future to adjust the array size dynamically based on the board type.

HyodaKazuaki replied:
> It might be a good idea to adjust the array size dynamically based on the board type.

Yes, I agree with your idea. Thank you for suggesting it.
Could you please implement this?

Yupopyoi commented:
@HyodaKazuaki 
Thank you for taking the time to review the PR.

I appreciate your kind words. And you're absolutely right — I made changes quite freely, so I completely understand that they ended up being breaking changes. I'm sorry for the inconvenience, and I agree that releasing them in the next major version makes the most sense.

PRを確認していただきありがとうございます。
ご指摘の通り、私が自由に書きすぎた結果、破壊的な変更になってしまいました。ご迷惑をおかけして申し訳ないです。

HyodaKazuaki commented:
Naming convention problem is can be addressed in a separate issue or it can certainly be handled within this PR.
I'll leave the decision to @Yupopyoi.

> If you'd prefer, I can revise the changes to keep backward compatibility. Just let me know.

Please let me know if you have any other ideas.


命名規則に関する問題であれば他issueとして対応できますし、もちろんこのPRで対応していただいてもOKです。
@Yupopyoi に判断をお任せします。

> 互換性を保つ方向で直すことも可能ですが、どうしましょうか？

他にアイデアがあればお教えいただきたいです。

Yupopyoi commented:
I'll rename the private variables and functions in this PR to follow the naming conventions.

I'd like to align on the C++ naming conventions.
Right now, I’m thinking of using:
- UpperCamelCase for function names
- lowerCamelCase for variables

Does that sound good to everyone? I know naming styles can vary between teams, so I’d like to make sure we’re aligned.

Also, for member variables, should we:
- use a m_ prefix?
- use a trailing underscore (_)?
- or leave them without any prefix/suffix?

ひとまず、外部に影響のない、private 変数・関数についてはこのPRで命名を変更していきますね。

命名規則については、すり合わせを行っておきたいです。
関数名はUpperCamelCase、変数名はlowerCamelCaseにするのが良いかと考えていますが、
このあたりは人によって流儀が分かれることもあると思うので、念のため確認させてください。

また、メンバ変数については、m_ のプレフィックスを使うか、アンダースコア _ を末尾に付けるか何も付けずにそのままにするかもあわせて決めておきたいです。

HyodaKazuaki commented:
According to [Arduino Style Guide for Creating Libraries](https://docs.arduino.cc/learn/contributions/arduino-library-style-guide/), we should use lower camel case for function names and the same for variables.

> Also, for member variables, should we:
>
>    use a m_ prefix?
>    use a trailing underscore (_)?
>    or leave them without any prefix/suffix?

Thank you for your suggestion!
Let me see, Use a trailing underscore for member variables.


[Arduino Style Guide for Creating Libraries](https://docs.arduino.cc/learn/contributions/arduino-library-style-guide/)によると、関数名や変数名にはlower camel caseを使うべきだと感じました。

> また、メンバ変数については、m_ のプレフィックスを使うか、アンダースコア _ を末尾に付けるか何も付けずにそのままにするかもあわせて決めておきたいです。

ご提案ありがとうございます。
メンバー変数には末尾にアンダースコアをつけてください。

Yupopyoi commented:
## Changes

### Naming Convention Standardization
- Renamed all function and variable names to **lowerCamelCase**
- Member variables now use a trailing underscore (e.g., `angle_`)
- Proportional gain `Kp` is **intentionally left unchanged** as it aligns with standard control notation

---

### `instances[]` Initialization
- `MAX_INTERRUPT_NUM` is now automatically set based on the target board (Uno / Leonardo / Mega)
- The size of `instances[]` adjusts accordingly

---

### Restored `rotate()` Function (for Compatibility)
- Restored `rotate()` as a **blocking legacy function**
- Not suitable for multi-servo or real-time applications
- Included for backward compatibility; safe to remove if no longer needed

---

### ISR Refactor: `updateAngleFromPWM()`
- Moved all angle computation logic out of `handleFeedback()` (ISR)
- Introduced `updateAngleFromPWM()` and now call it at the beginning of `update()`
- Results in better stability for multiple simultaneous feedback signals

---

Feedback is welcome. Let me know if anything should be adjusted.

---

## 変更点

いくつかのリファクタリングと互換性維持の対応を行いました：

---

### 命名規則の統一
- 関数名・変数名はすべて lowerCamelCase に変更しました。（Kpは例外）
- メンバー変数には `_` を末尾につけるスタイルに統一しました（例: `angle_`）。
- 比例ゲイン `Kp` については、一般的な制御工学の記法としてこの名前が自然と判断し、あえて変更していません。

---

### `instances[]` の定義改善
- 対応ボード（Uno / Leonardo / Mega）に応じて `MAX_INTERRUPT_NUM` が自動で定義されるようにし、それに応じて `instances[]` の要素数が決まるようにしました。

---

### `rotate()` 関数の復活（互換性対応）
- 旧コードとの互換性を保つため、同期的に動作する `rotate()` 関数を復活させました。
- ただし、ブロッキング処理であり、**複数サーボを制御する場合には非推奨**です。
- 残すか削除するかの判断はお任せします。

---

### ISRの負荷軽減 & 安定性向上
- ISR (`handleFeedback()`) 内の処理を最小化し、実際の角度演算を、新たに導入した `updateAngleFromPWM()` に分離しました。
- この関数は `update()` の先頭で呼び出されるようにしており、主処理と計測を分離したことで多モータ構成での安定性が向上しています。

---

不要な箇所や削除すべき箇所があれば、ご指摘いただければ対応します。
また、何か質問や気になる点があれば、遠慮なくコメントください。

Yupopyoi commented:
Sorry for the delayed reply, and thank you again for your thoughtful comments!

> I have a question. MAX_INTERRUPT_NUM is limited to 2 on boards other than Uno, Leonardo and Mega. Is this because all boards have two interrupt pins at least?

I haven't verified whether *all* boards have at least two interrupt pins.
For boards other than Uno, Leonardo, and Mega, I set MAX_INTERRUPT_NUM to 2 to avoid it being undefined.
This isn't necessarily the most appropriate value for every board — depending on the hardware, a larger or smaller number might be more suitable.
But for now, setting it to 2 seemed like a safer fallback than leaving it undefined.

> I'll add the deprecated attribute to this function later.

I agree with keeping rotate() as deprecated for compatibility.
Thank you for planning to add the deprecated attribute later — that sounds like a solid approach.

＞you can update Arduino Mega column to "supported" on SUPPORTED LIST

I'll go ahead and mark the Arduino Mega as "supported" in the SUPPORTED LIST in this PR.

---

返信が遅くなってしまい申し訳ありません。そして、今回も丁寧なコメントをありがとうございます。

> 1つ質問なのですが、UnoやLeonardo、Mega以外のボードでMAX_INTERRUPT_NUMが2に制限されているのは、すべてのボードで少なくとも2つの割り込みピンがあるからでしょうか？

すべてのボードに少なくとも2つの割り込みピンがあるかどうかは確認していません。
Uno、Leonardo、Mega以外のボードで未定義になるのを避けるため、暫定的に MAX_INTERRUPT_NUM を 2 に設定しました。
すべてのボードに対して最適な値とは限りませんし、ハードウェアによってはより大きい、あるいはより小さい値が適している可能性もあります。
ですが現時点では、未定義にするよりも2を指定する方が安全だと判断しました。

> 後で deprecated属性を関数につけておきたいと思います。

それが良い対応だと思います。

> SUPPORTED LISTのArduino Mega欄を"対応"に更新できます

このPR内で Arduino Mega を「対応」に更新しておきますね。

HyodaKazuaki commented:
@Yupopyoi The beta version library has been released!
You can try this from [GitHub Releases](https://github.com/HyodaKazuaki/Parallax-FeedBack-360-Servo-Control-Library-4-Arduino/releases/tag/v2.0.0-beta).

ベータ版のライブラリをリリースしました。
[GitHub Releases](https://github.com/HyodaKazuaki/Parallax-FeedBack-360-Servo-Control-Library-4-Arduino/releases/tag/v2.0.0-beta)からご確認いただけます。
